DBZ-3643 Fix to support NOBLOBs among binlog_row_image attributes.

Bue-von-hon opened the pull request:
resolves [DBZ-3643](https://issues.redhat.com/browse/DBZ-3643)

PTAL @gunnarmorling @Naros

Naros commented:
Hi @Bue-von-hon, thanks.  I commented on your Jira about the test case coverages, do you think you could add some test cases as I described in the Jira, thanks.

jpechane commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/strategy/AbstractConnectorConnection.java:
```diff
@@ -269,6 +269,29 @@ public boolean isBinlogRowImageFull() {
         }
     }
 
+    /**
+     * Determines whether the binlog format used by the database server is {@code binlog_row_image='NOBLOB'}.
+     *
+     * @return {@code true} if the {@code binlog_row_image} is set to {@code NOBLOB}, {@code false} otherwise
+     */
+    public boolean isBinlogRowImageNoblob() {
+        try {
+            final String rowImage = queryAndMap("SHOW GLOBAL VARIABLES LIKE 'binlog_row_image'", rs -> {
+                if (rs.next()) {
+                    return rs.getString(2);
+                }
+                // This setting was introduced in MySQL 5.6+ with default of 'FULL'.
+                // For older versions, assume 'FULL'.
+                return "FULL";
+            });
+            LOGGER.debug("binlog_row_image={}", rowImage);
+            return "NOBLOB".equalsIgnoreCase(rowImage);
+        }
+        catch (SQLException e) {
+            throw new DebeziumException("Unexpected error while connecting to the database and looking at BINLOG_ROW_IMAGE mode: ", e);
+        }
+    }
```
In fact, do we need to call `SHOW GLOBAL VARIABLES LIKE` twice? IMHO only one method is enough, just the name should be updated like `isBinlogRowImageFullOrNoBlob`

Bue-von-hon replied:
I've taken your feedback on board by integrating it into an existing method.
PTAL @Naros @jpechane

Bue-von-hon commented:
> Hi @Bue-von-hon, thanks. I commented on your Jira about the test case coverages, do you think you could add some test cases as I described in the Jira, thanks.

sure i will!

Bue-von-hon commented:
Thank @Naros and @jpechane for your patience and understanding.
I realized that the way to save the table schema is implemented by just saving the create query.
So I chose to modify the create query for the blob/text columns. 😅
If you have a better way, please let me know.

Bue-von-hon commented:
FYI
You need to modify line 192 of pom.xml inside debezium-connector-mysql like below to test.
<docker.dbs>debezium/mysql-server-test-binlog-row-image-database</docker.dbs>

p.s. I've added a new docker environment for testing, so please understand.

Bue-von-hon commented:
@Naros @jpechane gentle ping

jpechane commented on debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlConnectorNoblobT.java:
```diff
@@ -0,0 +1,114 @@
+ *
+ * @author Bue-Von-Hun
+ */
+public class MySqlConnectorNoblobT extends AbstractAsyncEngineConnectorTest {
```
```suggestion
public class MySqlConnectorNoBlobIT extends AbstractAsyncEngineConnectorTest {
```

jpechane commented:
@Bue-von-hon Could the whole testsuite pass when running against the new MySQL container image?
In general I am fine with adding a new one but it is important to have the test running as part of CI. In that case it is necessary to add it to `mysql-ci` profile so it is automatically executed on Github.

Bue-von-hon commented:
@Naros @jpechane I've resolved all the conflicts.
However, the usage of the UniqueDatabase class has changed, and it's taking some time to fix it. (Impact of DBZ-7693)
Previously, I instantiated the UniqueDatabase class directly with the new keyword, but now I can't do that.
So I changed the implementation to the MySqlUniqueDatabase class, which inherits from it, and now I get the error `"java.lang.IllegalStateException: java.sql.SQLException: Access denied for user 'mysqluser'@'localhost' (using password: YES)".`

I probably need to change the account information, but the MySqlUniqueDatabase class was just added, so I don't have a reference for it.
I'm still working on it, so please be patient.
Also, any additional information needed to troubleshoot the issue is always welcome.

Naros commented:
Hi @Bue-von-hon and @jpechane, I have a radical idea here, and I'm curious what you both think of this.  So my understanding here is that when `binlog_row_image=noblob`, this effectively tells the database to skip recording the blob and text columns in binlog events.  

Could we tackle this from the opposite side, rather than trying to manipulate the `CREATE TABLE` and attempting to omit the column at the outset, could we do this just before event emission using a `PostProcessor`? For example, the PostProcessor could examine the relational `Column` type and compare that with the binlog configs and if the `noblob` setting is enabled, any `TEXT` or `BLOB` column type would be dropped from the `Struct` just before we construct the `SourceRecord`.

jpechane commented:
@Naros Clever idea! Yes, let's do it this way.

Bue-von-hon commented:
Hi @Naros @jpechane leave information.
I'm trying to implement a way that if noblob mode is on when creating a SourceRecord, and the struct has a blob/text column, it will just ignore it.

It looks like there's going to be a major modification to the EventDispatcher class. (io.debezium.pipeline.EventDispatcher)
Below are the items that will be modified in EventDispatcher class:
- StreamingChangeRecordReceiver#changeRecord
- BufferingSnapshotChangeRecordReceiver#changeRecord
- IncrementalSnapshotChangeRecordReceiver#completeSnapshot

However, the event dispatcher class or method being changed must have the noblob configuration information.
I'm thinking of using sourceInfo inside the OffsetContext.

p.s. of course, inside sourceInfo there is no relevant information yet, I need to add it.

jpechane commented:
@Bue-von-hon Do you need to modify the `EventDispatcher`? Or would it be possible to make a subclass in MySQL/MariaDB and add the modifications there?

Naros commented:
Hi @Bue-von-hon it appears there are some test failures, could you take a look?

Naros commented:
Hi @Bue-von-hon can you remove the merge commit please and prefix the commits with DBZ-3643?

jpechane commented:
@Bue-von-hon Could you please run `-Passembly` profile locally to get the code formatted and packages built so the tests are then executed on CI?

Bue-von-hon commented:
@Naros @jpechane 
I tried to drop BLOB/TEXT columns using the SchemaBuilder class.
However, an exception is occurring due to a mismatch with the schema stored in the DataCollectionSchema class.
Instead of dropping the columns, how about inserting null values instead?

Bue-von-hon commented:
@jpechane @Naros 
Do all co-authors need to sign off as well?
I couldn’t find this mentioned in the blog post, so I’m leaving a question here.

Bue-von-hon commented:
It seems the build is broken. I’ll address the issue and request a review again once it’s fixed.

Bue-von-hon committed:
DBZ-3643 Support MySQL binlog_row_image=NOBLOB mode

When MySQL runs with binlog_row_image=NOBLOB, unchanged BLOB/TEXT columns
are omitted from binlog row events. This adds a MySQL-specific
MysqlEventDispatcher that substitutes the configured unavailable-value
placeholder for those columns so downstream consumers receive a consistent
record shape, mirroring the TOAST handling used by the Postgres and Oracle
connectors. Includes a dedicated NOBLOB integration test and a
binlog_row_image=NOBLOB test server image.

Co-authored-by: chungeun-choi <cucuridas@gmail.com>
Co-authored-by: sean-k1 <uhs2000@naver.com>
Signed-off-by: Bue-von-hon <dkssudvn2@gmail.com>
Signed-off-by: chungeun-choi <cucuridas@gmail.com>
Signed-off-by: sean-k1 <uhs2000@naver.com>

Bue-von-hon committed:
DBZ-3643 Fix MysqlEventDispatcher generic type to bind TableId

MysqlEventDispatcher declared a second free type parameter and passed a
MySqlDatabaseSchema (a DatabaseSchema<TableId>) to the EventDispatcher
super constructor that expects DatabaseSchema<T>, which failed to compile
with "no suitable constructor found". Bind the dispatcher directly to
TableId by dropping the second type parameter, remove the now-redundant
cast and unused import, and update the call sites in MySqlConnectorTask
and MySqlChangeEventSourceFactory accordingly.

Signed-off-by: Bue-von-hon <dkssudvn2@gmail.com>

Bue-von-hon committed:
DBZ-3643 Use JUnit 5 annotations in MySqlConnectorNoBlobIT

The test imported JUnit 4 annotations (org.junit.Before/After/Test) while
the debezium-connector-mysql module only provides JUnit 5 on the test
classpath, breaking testCompile. Switch to org.junit.jupiter.api
(@BeforeEach/@AfterEach/@Test) to match the rest of the module.

Signed-off-by: Bue-von-hon <dkssudvn2@gmail.com>

Bue-von-hon commented:
FYI.
```
[ERROR] Tests run: 2, Failures: 1, Errors: 1, Skipped: 0, Time elapsed: 12.05 s <<< FAILURE! -- in io.debezium.connector.mysql.MySqlConnectorNoBlobIT
[ERROR] io.debezium.connector.mysql.MySqlConnectorNoBlobIT.textAndBlobColumnShouldNotBeContainedInBeforeAndAfterSectionsDuringUpdate -- Time elapsed: 11.68 s <<< FAILURE!
java.lang.AssertionError:

Expecting actual:
  0
to be greater than or equal to:
  3

        at io.debezium.connector.mysql.MySqlConnectorNoBlobIT.textAndBlobColumnShouldNotBeContainedInBeforeAndAfterSectionsDuringUpdate(MySqlConnectorNoBlobIT.java:154)

[ERROR] io.debezium.connector.mysql.MySqlConnectorNoBlobIT.textAndBlobColumnShouldNotBeContainedDuringSnapshot -- Time elapsed: 0.305 s <<< ERROR!
java.lang.IllegalStateException: java.sql.SQLException: Can't create database 'connector_noblob_mode_test_1qryu6g'; database exists
        at io.debezium.connector.binlog.util.UniqueDatabase.createAndInitialize(UniqueDatabase.java:182)
        at io.debezium.connector.binlog.util.UniqueDatabase.createAndInitialize(UniqueDatabase.java:148)
        at io.debezium.connector.mysql.MySqlConnectorNoBlobIT.beforeEach(MySqlConnectorNoBlobIT.java:64)
Caused by: java.sql.SQLException: Can't create database 'connector_noblob_mode_test_1qryu6g'; database exists
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:121)
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:114)
        at com.mysql.cj.jdbc.StatementImpl.executeInternal(StatementImpl.java:837)
        at com.mysql.cj.jdbc.StatementImpl.execute(StatementImpl.java:685)
        at io.debezium.jdbc.JdbcConnection.lambda$execute$3(JdbcConnection.java:437)
        at io.debezium.jdbc.JdbcConnection.execute(JdbcConnection.java:456)
        at io.debezium.jdbc.JdbcConnection.execute(JdbcConnection.java:431)
        at io.debezium.connector.binlog.util.UniqueDatabase.createAndInitialize(UniqueDatabase.java:178)
        ... 2 more

[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   MySqlConnectorNoBlobIT.textAndBlobColumnShouldNotBeContainedInBeforeAndAfterSectionsDuringUpdate:154
Expecting actual:
  0
to be greater than or equal to:
  3

[ERROR] Errors:
[ERROR]   MySqlConnectorNoBlobIT.beforeEach:64 » IllegalState java.sql.SQLException: Can't create database 'connector_noblob_mode_test_1qryu6g'; database exists
[INFO]
[ERROR] Tests run: 2, Failures: 1, Errors: 1, Skipped: 0
[INFO]
[INFO]
[INFO] --- docker:0.48.1:stop (stop) @ debezium-connector-mysql ---
02:11:09.445 mysql2026-06-27T02:11:09.445231Z 0 [System] [MY-013172] [Server] Received SHUTDOWN from user <via user signal>. Shutting down mysqld (Version: 8.2.0).
02:11:17.949 mysql2026-06-27T02:11:17.949760Z 0 [System] [MY-010910] [Server] /usr/sbin/mysqld: Shutdown complete (mysqld 8.2.0)  MySQL Community Server - GPL.
02:11:17.956 mysql2026-06-27T02:11:17.956213Z 0 [System] [MY-015016] [Server] MySQL Server - end.
[INFO] DOCKER> [debezium/mysql-server-test-binlog-row-image-database:latest]: Stop and removed container 030d3cb52ab0 after 0 ms
[INFO]
[INFO] --- failsafe:3.5.5:verify (verify) @ debezium-connector-mysql ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:10 min
[INFO] Finished at: 2026-06-27T11:11:18+09:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-failsafe-plugin:3.5.5:verify (verify) on project debezium-connector-mysql: There are test failures.
[ERROR]
[ERROR] See /Users/taehun.kim/workspace/debezium/debezium-connector-mysql/target/failsafe-reports for the individual test results.
[ERROR] See dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

Naros commented:
Hi @Bue-von-hon the problem is you are not creating the `DATABASE` correctly, you need to use the `TestHelper` rather than creating the instance directly. Please follow the design pattern found [here](https://github.com/debezium/debezium/blob/22f8f2c7af13e02684e694258f30794ebf574c3f/debezium-connector-binlog/src/test/java/io/debezium/connector/binlog/BinlogBinaryModeIT.java#L38-L39) and that should solve your issue.

Bue-von-hon committed:
DBZ-3643 Use TestHelper to create UniqueDatabase in MySqlConnectorNoBlobIT

Creating the UniqueDatabase directly as a static anonymous subclass reused
one random identifier for the whole class, so the second test method failed
in beforeEach with 'Can't create database ...; database exists'. Obtaining
the database from TestHelper via the UniqueDatabaseProvider service loader
follows the established test design pattern, creates a uniquely named
database per test instance and inherits the MySQL-specific JDBC, SSL and
timezone configuration from MySqlUniqueDatabase.

Also add the mysql-binlog-row-image Maven profile that runs this test
against the dedicated binlog_row_image=NOBLOB Docker image. The test is
excluded from the default failsafe run because its assertions only hold
for a server running with binlog_row_image=NOBLOB.

Co-authored-by: chungeun-choi <cucuridas@gmail.com>
Co-authored-by: sean-k1 <uhs2000@naver.com>
Signed-off-by: Bue-von-hon <dkssudvn2@gmail.com>
Signed-off-by: chungeun-choi <cucuridas@gmail.com>
Signed-off-by: sean-k1 <uhs2000@naver.com>

Bue-von-hon committed:
DBZ-3643 Handle NOBLOB row images in binlog streaming

When MySQL runs with binlog_row_image=NOBLOB, row images in UPDATE and
DELETE events omit unchanged BLOB/TEXT columns, so the deserialized row
arrays are narrower than the table schema and the streaming source
rejected them with a column-count mismatch, stopping the connector.

Expand each row image to the full table width using the event's
included-columns bitmap and substitute the configured unavailable-value
placeholder for the omitted BLOB/TEXT columns, mirroring the TOAST
handling of the Postgres connector. FULL row images are returned
unchanged. The BLOB/TEXT column check is extracted to a shared
RowImageUtils so the MySQL snapshot dispatcher reuses it.

Rewrite the NOBLOB update test to expect the fixed-width value schema
with placeholder values and add coverage for DELETE before images.

Co-authored-by: chungeun-choi <cucuridas@gmail.com>
Co-authored-by: sean-k1 <uhs2000@naver.com>
Signed-off-by: Bue-von-hon <dkssudvn2@gmail.com>
Signed-off-by: chungeun-choi <cucuridas@gmail.com>
Signed-off-by: sean-k1 <uhs2000@naver.com>

Naros commented:
Hi @Bue-von-hon I've collected all the commits from here and aligned the changes. It's now superseded in the above mentioned PR. I'm going to close this one in favor of the new one that is rebased. I kept yours and all the other attributions on the commits.
